File: Solution with Hash Map.java

import java.util.Map;
import java.util.HashMap;

public class Solution {

    public int findPairs(int[] nums, int k) {

        Map<Integer, Integer> frequencyMap = new HashMap<>();
        for (int n : nums) {
            frequencyMap.put(n, frequencyMap.getOrDefault(n, 0) + 1);
        }

        int countPairs = 0;
        for (int n : frequencyMap.keySet()) {
            if (isValidPair(k, n, frequencyMap)) {
                countPairs++;
            }
        }
        return countPairs;
    }

    public boolean isValidPair(int k, int n, Map<Integer, Integer> frequencyMap) {
        return frequencyMap.containsKey(k + n) && (k > 0 || (frequencyMap.get(k + n) > 1));
    }
}

File: Solution with Hash Map.cpp

#include<vector>
#include<unordered_map>
using namespace std;

class Solution {
    
public:

    int findPairs(vector<int>& nums, int k) {

        unordered_map<int, int> frequencyMap;
        for (int n : nums) {
            frequencyMap[n]++;
        }

        int countPairs = 0;
        for (const auto& n : frequencyMap) {
            if (isValidPair(k, n.first, frequencyMap)) {
                countPairs++;
            }
        }
        return countPairs;
    }

    bool isValidPair(int k, int n, unordered_map<int, int>& frequencyMap) {
        return frequencyMap.find(k + n) != frequencyMap.end() && (k > 0 || (frequencyMap[k + n] > 1));
    }
};
